src/java.base/share/classes/jdk/internal/ref/CleanerImpl.java

```diff
@@ -231,4 +223,135 @@ protected void performCleanup() {
             // no action
         }
     }
+
+    /**
+     * A specialized implementation that tracks phantom cleanables.
+     */
+    static final class CleanableList {
+        /**
+         * Capacity for a single node in the list.
+         * This balances memory overheads vs locality vs GC walking costs.
+         */
+        static final int NODE_CAPACITY = 4096;
+
+        /**
+         * Head node. This is the only node where PhantomCleanables are
+         * added to or removed from. This is the only node with variable size,
+         * all other nodes linked from the head are always at full capacity.
+         */
+        private Node head;
+
+        /**
+         * Cached node instance to provide better behavior near NODE_CAPACITY
+         * threshold: if list size flips around NODE_CAPACITY, it would reuse
+         * the cached node instead of wasting and re-allocating a new node all
+         * the time.
+         */
+        private Node cache;
+
+        public CleanableList() {
+            reset();
+        }
+
+        /**
+         * Testing support: reset list to initial state.
+         */
+        synchronized void reset() {
+            this.head = new Node();
+        }
+
+        /**
+         * Returns true if cleanable list is empty.
+         *
+         * @return true if the list is empty
+         */
+        public synchronized boolean isEmpty() {
+            // Head node size is zero only when the entire list is empty.
+            return head.size == 0;
+        }
+
+        /**
+         * Insert this PhantomCleanable in the list.
+         */
+        public synchronized void insert(PhantomCleanable<?> phc) {
+            if (head.size == NODE_CAPACITY) {
```

Comment:
Given that `head.size` is a mutable member, it will be clearer to start out reading the current head.size into a local final variable and use that variable throughout the method where head.size is read.

Reply:
I don't quite agree on this one: in `insert`, `head.size` is not stable, as we can switch the `head` in the middle of the method, when we insert another head. Keeping `head.size` in a local variable raises a question whether that local is out of sync. The only place I can see this _might_ make sense is saving `head.size` on two adjacent lines after `head` is indeed stable.

Overall, I think we are in nitpicking territory, and this is 15-th revision of this PR. I would prefer us not to do any changes that are not substantial.